vip-woocommerce/action-scheduler.php

```diff
@@ -0,0 +1,166 @@
+
+	// Safety cap.
+	const MAX_ALLOWED_DYNAMIC_QUEUES = 15;
+
+	public function init() {
+		if ( ! $this->is_enabled() ) {
+			return;
+		}
+
+		// Disable the async (frontend ajax) queue. We only want to run AS in cron.
+		add_filter( 'action_scheduler_allow_async_request_runner', '__return_false', 30 );
+
+		// Allow each queue to run for 120 seconds by default.
+		add_filter( 'action_scheduler_queue_runner_time_limit', fn() => 120, 30 );
+
+		// Configure concurrency.
+		add_filter( 'a8c_cron_control_concurrent_event_whitelist', [ $this, 'configure_cron_control_concurrency' ] );
+		add_filter( 'action_scheduler_queue_runner_concurrent_batches', [ $this, 'configure_action_scheduler_concurrency' ], 30 );
+
+		add_action( self::QUEUE_PROCESSOR_CRON_EVENT, [ $this, 'process_queue' ] );
+
+		if ( wp_doing_cron() || ( is_admin() && ! wp_doing_ajax() ) ) {
+			add_action( 'shutdown', [ $this, 'maybe_dispatch_new_queues' ] );
+		}
+	}
+
+	private function is_enabled() {
+		$dependencies_exist = class_exists( 'WooCommerce' ) && class_exists( 'ActionScheduler' ) && class_exists( '\Automattic\WP\Cron_Control\Events_Store' );
+		return apply_filters( 'vip_action_schedule_dynamic_queue_enabled', $dependencies_exist );
+	}
+
+	/**
+	 * Increase allowed event concurrency in cron control.
+	 * Note: This needs to happen before the after_setup_theme hook, priority 10.
+	 */
+	public function configure_cron_control_concurrency( $whitelist ) {
+		$whitelist[ self::QUEUE_PROCESSOR_CRON_EVENT ] = $this->get_max_allowed_queue_jobs();
+		return $whitelist;
+	}
+
+	/**
+	 * Increase allowed batch concurrency in action scheduler.
+	 */
+	public function configure_action_scheduler_concurrency() {
+		// Increment by 1 to account for the default `action_scheduler_run_queue` recurring action.
+		return $this->get_max_allowed_queue_jobs() + 1;
+	}
+
+	private function get_max_allowed_queue_jobs() {
+		$total_cron_control_concurrency_limit = defined( '\Automattic\WP\Cron_Control\JOB_CONCURRENCY_LIMIT' ) ? \Automattic\WP\Cron_Control\JOB_CONCURRENCY_LIMIT : 10;
+
+		// Allow up to 33% of the cron-control queue to be used for action scheduler processing.
+		return (int) min( ceil( $total_cron_control_concurrency_limit / 3 ), self::MAX_ALLOWED_DYNAMIC_QUEUES );
+	}
+
+	private function get_queue_timeout_limit() {
+		// Note: We set this to 120 seconds in a filter above.
+		return absint( apply_filters( 'action_scheduler_queue_runner_time_limit', 30 ) );
+	}
+
+	/**
+	 * Every few minutes, we determine if we should dispatch additional AS queues.
+	 */
+	public function maybe_dispatch_new_queues() {
+		$dispatch_interval = max( $this->get_queue_timeout_limit() - 10, 60 );
+		if ( ! wp_cache_add( 'dynamic-queue-scheduler-lock', 'locked', 'vip', $dispatch_interval ) ) {
+			// Only dispatch new queues around the time it takes to finish the previous round, or once a minute minimum.
+			return;
+		}
+
+		$pending_cron_jobs_count = $this->get_pending_queue_job_count();
```

Comment:
So this `$pending_cron_jobs_count` is the number of queues as in Action Scheduler events that each will process jobs, correct? This naming is really hard to get right I think. It is easy to confuse it with the jobs that will be processed by the queue. 

What do you think about naming it `$current_queue_count` ?

Reply:
Yeah, lots of shared wording.

Cron job = `vip_action_scheduler_run_queue` event
Each job starts up 1 AS queue.
Each AS queue processes actions.

____

I think the important factor to convey here though is that they are scheduled cron jobs that have yet to run. So it's not technically an AS queue yet, just pending cron jobs.

Reply:
Added some extra documentation in the class docblock to help clarify this process.